File: proj2md-py/proj2md/render.py
"""proj2md Markdown 渲染：目录树 / 索引表 / 文件正文 / 附录。"""
from __future__ import annotations
from datetime import datetime

from .defaults import MARK, MARK_RE, TOOL, VERSION
from .i18n import t
from .util import estimate_tokens, fence_for, fence_lang_of, fmt_size, md_code_span, md_slug, md_table_cell

# ─────────────────────────── 渲染（Markdown） ───────────────────────────
def build_tree(records, root_label: str) -> str:
    tree = {}
    for r in records:
        node = tree
        parts = list(r.rel.parts)
        for part in parts[:-1]:
            node = node.setdefault(part, {})
        node[parts[-1]] = None
    lines = [root_label + "/"]
    def walk(node, prefix):
        items = sorted(node.items(), key=lambda kv: (kv[1] is None, kv[0].lower()))
        for i, (name, child) in enumerate(items):
            last = (i == len(items) - 1)
            lines.append(prefix + ("└── " if last else "├── ") + name + ("/" if child else ""))
            if child:
                walk(child, prefix + ("    " if last else "│   "))
    walk(tree, "")
    return "\n".join(lines) + "\n"
def render(cfg: Config, records, skipped, prompt_text: str, root_name: str,
           part_label: str = "", pruned_hidden=None) -> str:
    n = len(records)
    tot_lines = sum(r.lines for r in records)
    tot_size = sum(len(r.content.encode("utf-8")) for r in records)  # ★ 修复（P0-1）：字节口径
               # （原按字符数计，与控制台报告 len(text.encode("utf-8")) 口径不一致）
    tot_tokens = sum(estimate_tokens(r.content) for r in records)
    now = datetime.now().strftime("%Y-%m-%d %H:%M")
    pruned_hidden = pruned_hidden or []
    # ── 头部各节：标题/说明/需求/目录树 ──
    head = []
    meta = [
        "- " + t("doc_meta_time", now=now),
        "- " + t("doc_meta_project", root=root_name),
        "- " + (t("doc_meta_files_skipped", n=n, skipped=len(skipped)) if skipped
                else t("doc_meta_files", n=n)),
        "- " + t("doc_meta_lines", lines=f"{tot_lines:,}"),
        "- " + t("doc_meta_size", size=fmt_size(tot_size)),  # ★
        "- " + t("doc_meta_tokens", tokens=f"{tot_tokens:,}"),
    ]
    head.append(t("doc_title", root=root_name, label=part_label) + "\n\n" + "\n".join(meta))
    if cfg.ai_header:
        head.append(t("doc_ai_header") + t("ai_notes", root=root_name))
    if prompt_text:
        head.append(t("doc_prompt") + prompt_text)
    if cfg.show_tree:
        tree = build_tree(records, root_name).rstrip("\n")
        tf = fence_for(tree)
        head.append(t("doc_tree") + tf + "\n" + tree + "\n" + tf)
    # ── 文件正文节：### 序号. 路径 + 围栏代码块 ──
    file_secs = []
    for i, r in enumerate(records, 1):
        path = r.rel.as_posix()
        content = r.content
        if cfg.line_numbers:
            ls = content.split("\n")
            if ls and ls[-1] == "":
                ls.pop()
            content = "\n".join(f"{k:>5} | {ln}" for k, ln in enumerate(ls, 1)) + "\n"
        f = fence_for(content)
        info = [f"`{r.language}`", t("doc_lines_unit", n=r.lines)]
        if r.encoding and not r.encoding.startswith("utf"):
            info.append(t("doc_encoding", enc=r.encoding))
        if r.truncated:
            info.append(t("doc_truncated"))
        # 围栏行末尾加内部标记，稍后用于回填索引中的起始行号
        fence_line = f + fence_lang_of(r.rel) + MARK + str(i) + MARK
        file_secs.append(
            f"### {i}. {path}\n\n"
            f"**{i}/{n}** · " + " · ".join(info) + "\n\n" +
            fence_line + "\n" + content + f)
    # ── 索引表（起始行先占位，组装后按标记回填真实行号） ──
    def make_index(starts):
        rows = [t("doc_index_cols"), "|---:|:---|:---|---:|---:|"]
        for i, (r, s) in enumerate(zip(records, starts), 1):
            path = r.rel.as_posix()
            cell = md_table_cell(path)
            anchor = md_slug(f"{i}. {path}")
            rows.append(f"| {i} | [{cell}](#{anchor}) | {r.language} | {r.lines} | {s} |")
        return t("doc_index") + "\n".join(rows)
    # ── 尾部各节：附录 / 结尾 ──
    tail = []
    if file_secs:
        tail.append(t("doc_source") + "\n\n".join(file_secs))
    if skipped:
        items = [t("doc_skip_item", path=md_code_span(rel), reason=reason)
                 for rel, reason in skipped[:50]]
        if len(skipped) > 50:
            items.append(t("doc_more_skipped", n=len(skipped) - 50))
        tail.append(t("doc_appendix_skipped") + "\n".join(items))
    if pruned_hidden:
        shown = pruned_hidden[:30]
        items = [t("doc_hidden_item", path=md_code_span(d)) for d in shown]
        if len(pruned_hidden) > 30:
            items.append(t("doc_more_hidden", n=len(pruned_hidden) - 30))
        tail.append(t("doc_appendix_hidden") + "\n".join(items))
    tail.append(t("doc_end", n=n, lines=f"{tot_lines:,}", tokens=f"{tot_tokens:,}",
                  tool=TOOL, ver=VERSION, now=now))
    dummy_index = make_index([0] * n) if (cfg.show_index and n) else None
    secs = head + ([dummy_index] if dummy_index else []) + tail
    doc = "\n\n".join(secs) + "\n"
    # ── 回填真实起始行号，并整体移除内部标记（标记+序号一起删除） ──
    if dummy_index:
        starts = [0] * n
        for li, ln in enumerate(doc.split("\n"), 1):
            m = MARK_RE.search(ln)
            if m:
                starts[int(m.group(1)) - 1] = li + 1   # 围栏行的下一行即正文首行
        doc = MARK_RE.sub("", doc)
        if all(s > 0 for s in starts):
            doc = doc.replace(dummy_index, make_index(starts), 1)
        else:
            doc = MARK_RE.sub("", doc)
    return doc

File: proj2md-js/lib/render.js
'use strict';
const D = require('./defaults.js');
const { t } = require('./i18n.js');
const {
  estimateTokens, fmtSize, fmtInt,
  fenceFor, fenceLangOf, mdSlug, mdCodeSpan, mdTableCell,
} = require('./util.js');
const MARK = D.MARK;
const MARK_SEARCH_RE = /\x00(\d+)\x00/;
const MARK_STRIP_RE = /\x00\d+\x00/g;
function localNow() {
  const d = new Date();
  const p = (x) => String(x).padStart(2, '0');
  return `${d.getFullYear()}-${p(d.getMonth() + 1)}-${p(d.getDate())} ${p(d.getHours())}:${p(d.getMinutes())}`;
}
function replaceOnce(hay, needle, replFn) {
  const idx = hay.indexOf(needle);
  if (idx < 0) return hay;
  return hay.slice(0, idx) + replFn() + hay.slice(idx + needle.length);
}
function buildTree(records, rootLabel) {
  const tree = {};
  for (const r of records) {
    let node = tree;
    const parts = r.rel.split('/');
    for (let i = 0; i < parts.length - 1; i++) {
      if (!node[parts[i]] || typeof node[parts[i]] !== 'object') node[parts[i]] = {};
      node = node[parts[i]];
    }
    node[parts[parts.length - 1]] = null;
  }
  const lines = [rootLabel + '/'];
  const walk = (node, prefix) => {
    const items = Object.entries(node).sort((a, b) => {
      const fa = a[1] === null ? 1 : 0, fb = b[1] === null ? 1 : 0; // 目录在前，文件在后
      if (fa !== fb) return fa - fb;
      const la = a[0].toLowerCase(), lb = b[0].toLowerCase();
      return la < lb ? -1 : la > lb ? 1 : 0;
    });
    items.forEach(([name, child], i) => {
      const last = i === items.length - 1;
      lines.push(prefix + (last ? '└── ' : '├── ') + name + (child ? '/' : ''));
      if (child) walk(child, prefix + (last ? '    ' : '│   '));
    });
  };
  walk(tree, '');
  return lines.join('\n') + '\n';
}
function render(cfg, records, skipped, promptText, rootName, partLabel = '', prunedHidden = null) {
  const n = records.length;
  const totLines = records.reduce((s, r) => s + r.lines, 0);
  const totSize = records.reduce((s, r) => s + Buffer.byteLength(r.content, 'utf8'), 0); // ★ 修复（P0-1）：字节口径（原为字符数，与控制台报告不一致）
  const totTokens = records.reduce((s, r) => s + estimateTokens(r.content), 0);
  const now = localNow();
  prunedHidden = prunedHidden || [];
  // ── 头部各节：标题/说明/需求/目录树 ──
  const head = [];
  const meta = [
    '- ' + t('doc_meta_time', { now }),
    '- ' + t('doc_meta_project', { root: rootName }),
    '- ' + (skipped.length
      ? t('doc_meta_files_skipped', { n, skipped: skipped.length })
      : t('doc_meta_files', { n })),
    '- ' + t('doc_meta_lines', { lines: fmtInt(totLines) }),
    '- ' + t('doc_meta_size', { size: fmtSize(totSize) }), // ★
    '- ' + t('doc_meta_tokens', { tokens: fmtInt(totTokens) }),
  ];
  head.push(t('doc_title', { root: rootName, label: partLabel }) + '\n\n' + meta.join('\n'));
  if (cfg.aiHeader) head.push(t('doc_ai_header') + t('ai_notes', { root: rootName }));
  if (promptText) head.push(t('doc_prompt') + promptText);
  if (cfg.showTree) {
    const tree = buildTree(records, rootName).replace(/\n$/, '');
    const tf = fenceFor(tree);
    head.push(t('doc_tree') + tf + '\n' + tree + '\n' + tf);
  }
  // ── 文件正文节 ──
  const fileSecs = [];
  records.forEach((r, idx) => {
    const i = idx + 1;
    let content = r.content;
    if (cfg.lineNumbers) {
      const ls = content.split('\n');
      if (ls.length && ls[ls.length - 1] === '') ls.pop();
      content = ls.map((ln, k) => String(k + 1).padStart(5) + ' | ' + ln).join('\n') + '\n';
    }
    const f = fenceFor(content);
    const info = ['`' + r.language + '`', t('doc_lines_unit', { n: r.lines })];
    if (r.encoding && !r.encoding.startsWith('utf')) info.push(t('doc_encoding', { enc: r.encoding }));
    if (r.truncated) info.push(t('doc_truncated'));
    // 围栏行末尾加内部标记，稍后用于回填索引中的起始行号
    const fenceLine = f + fenceLangOf(r.rel) + MARK + i + MARK;
    fileSecs.push(
      '### ' + i + '. ' + r.rel + '\n\n' +
      '**' + i + '/' + n + '** · ' + info.join(' · ') + '\n\n' +
      fenceLine + '\n' + content + f
    );
  });
  // ── 索引表（起始行先占位，组装后按标记回填真实行号）──
  const makeIndex = (starts) => {
    const rows = [t('doc_index_cols'), '|---:|:---|:---|---:|---:|'];
    records.forEach((r, idx) => {
      const i = idx + 1;
      const cell = mdTableCell(r.rel);
      const anchor = mdSlug(i + '. ' + r.rel);
      rows.push('| ' + i + ' | [' + cell + '](#' + anchor + ') | ' + r.language + ' | ' + r.lines + ' | ' + starts[idx] + ' |');
    });
    return t('doc_index') + rows.join('\n');
  };
  // ── 尾部各节 ──
  const tail = [];
  if (fileSecs.length) tail.push(t('doc_source') + fileSecs.join('\n\n'));
  if (skipped.length) {
    const items = skipped.slice(0, 50).map(([rel, reason]) => t('doc_skip_item', { path: mdCodeSpan(rel), reason }));
    if (skipped.length > 50) items.push(t('doc_more_skipped', { n: skipped.length - 50 }));
    tail.push(t('doc_appendix_skipped') + items.join('\n'));
  }
  if (prunedHidden.length) {
    const items = prunedHidden.slice(0, 30).map((d) => t('doc_hidden_item', { path: mdCodeSpan(d) }));
    if (prunedHidden.length > 30) items.push(t('doc_more_hidden', { n: prunedHidden.length - 30 }));
    tail.push(t('doc_appendix_hidden') + items.join('\n'));
  }
  tail.push(t('doc_end', {
    n, lines: fmtInt(totLines), tokens: fmtInt(totTokens),
    tool: D.TOOL, ver: D.VERSION, now,
  }));
  const dummyIndex = (cfg.showIndex && n) ? makeIndex(records.map(() => 0)) : null;
  const secs = head.concat(dummyIndex ? [dummyIndex] : [], tail);
  let doc = secs.join('\n\n') + '\n';
  // ── 回填真实起始行号，并整体移除内部标记 ──
  if (dummyIndex) {
    const starts = new Array(n).fill(0);
    const lines = doc.split('\n');
    for (let li = 0; li < lines.length; li++) {
      const m = MARK_SEARCH_RE.exec(lines[li]);
      if (m) starts[Number(m[1]) - 1] = li + 2; // 围栏行的下一行即正文首行
    }
    doc = doc.replace(MARK_STRIP_RE, '');
    if (starts.every((s) => s > 0)) {
      doc = replaceOnce(doc, dummyIndex, () => makeIndex(starts));
    }
  }
  return doc;
}
module.exports = { buildTree, render };

